lighthouse-core/audits/byte-efficiency/unused-javascript.js

```diff
@@ -0,0 +1,184 @@
+    return {
+      used: objA.used + objB.used,
+      unused: objA.unused + objB.unused,
+    };
+  }
+
+  /**
+   * Determines the length of unused and used portions in a set of ranges.
+   * Because the protocol reports usage for each function and ranges within each function, nested
+   * regions are represented numerous times. This function makes sure to only count each string
+   * section once. See unit and smoke tests for examples.
+   *
+   * @param {!Array<{isUsed: boolean, startOffset: number, endOffset: number
+   *    used: number, unused: number}>} ranges
+   * @return {{used: number, unused: number}}
+   */
+  static findUsedAndUnused(ranges) {
+    if (ranges.length === 0) {
+      return {used: 0, unused: 0};
+    } else if (ranges.length === 1) {
+      return ranges[0];
+    }
+
+    const firstRange = ranges[0];
+    const secondRange = ranges[1];
+    if (secondRange.startOffset >= firstRange.endOffset) {
+      // Ranges aren't nested, we can simply continue on to the rest.
+      return this._add(
+        this._add(firstRange, {used: secondRange.startOffset - firstRange.endOffset, unused: 0}),
+        this.findUsedAndUnused(ranges.slice(1))
+      );
+    } else {
+      let endIndex = 1;
+      while (endIndex < ranges.length && ranges[endIndex].startOffset < firstRange.endOffset) {
+        endIndex++;
+      }
+
+      const spanningRanges = ranges.slice(1, endIndex);
+      const restOfRanges = ranges.slice(endIndex);
+      if (firstRange.isUsed) {
+        // Nested ranges may potentially be unused which we need to account for.
+        const earliestStart = secondRange.startOffset;
+        const latestEnd = spanningRanges.reduce((max, next) => {
+          return max.endOffset >= next.endOffset ? max : next;
+        }).endOffset;
+        const nextStart = (restOfRanges[0] && restOfRanges[0].startOffset) || firstRange.endOffset;
+
+        const leftEdge = {used: earliestStart - firstRange.startOffset, unused: 0};
+        const rightEdge = {used: nextStart - latestEnd, unused: 0};
+        const inner = this.findUsedAndUnused(spanningRanges);
+        const rest = this.findUsedAndUnused(restOfRanges);
+        return this._add(this._add(leftEdge, rightEdge), this._add(inner, rest));
+      } else {
+        // Nested ranges won't be used if the entire parent wasn't used, just skip the children.
+        return this._add(firstRange, this.findUsedAndUnused(restOfRanges));
+      }
+    }
+  }
+
+  /**
+   * @param {!Object} script
+   * @param {{transferSize: number}} networkRecord
+   */
+  static computeWaste(script, networkRecord) {
+    const url = URL.getDisplayName(script.url);
+
+    let numUnusedFunctions = 0;
+    const ranges = [];
+    script.functions.forEach(func => {
+      let functionIsUsed = false;
+      func.ranges.forEach(range => {
+        const isUsed = range.count > 0;
+        const span = range.endOffset - range.startOffset;
+        const used = isUsed ? span : 0;
```

Comment:
what are these offsets? Lines? Character position? A better name for `used` and `unused` would be good here to hint at that, as they sound boolean at first. `usedLength`? `usedLines`?